### ISSUE TITLE
Playing from Capture Card or Webcam, Video device not found even though listed

### ISSUE THREAD

**Issue by justin-pierce** [mpv-player/mpv#8818]
### Important Information

Provide following Information:
- mpv 0.33.1
- macOS 11.1
- Source of the mpv binary or bundle: macOS builds by stolendata
- If known which version of mpv introduced the problem: unknown

### Reproduction steps

Run in terminal: 
```
mpv av://avfoundation:video="HD Pro Webcam C920"
```

"HD Pro Webcam C920" is derived from 

```
ffmpeg -f avfoundation -list_devices true -i ""
```

Also tried ShadowCast (capture card), same thing. Goal is to achieve [this](https://www.reddit.com/r/GenkiLab/comments/mzw9a2/lower_latency_option_for_windows_mpvio/) on MacOS. 

### Expected behavior

mpv opens with view of webcam feed

### Actual behavior
```
[ffmpeg] AVFoundation indev: Video device not found
[lavf] avformat_open_input() failed
Failed to recognize file format.
```

### Log file

[output.txt](https://github.com/mpv-player/mpv/files/6442903/output.txt)

When I try to capture the feed directly in ffmpeg, I have to specify a framerate -- is there a way to do that in mpv? The fps flag didn't seem to work -- thinking maybe it instead needs to be in the ffmpeg args?

**Comment by Akemi** [comment]
you can try `av://avfoundation:0:none`, eg instead of using the name using the index. the `:none` is to select no audio device.

you can specify various demuxer properties via `demuxer-lavf-o`, some examples:

```
--demuxer-lavf-o=video_size=1920x1080 
--demuxer-lavf-o=pixel_format=uyvy420
--demuxer-lavf-o=framerate=23.976
```

**Comment by justin-pierce** [comment]
Removing the video/audio tags seems to help, along with specifying framerate. 

```
V: 00:00:09 / 00:00:10 (97%)
[lavf] error reading packet: Resource temporarily unavailable.
[lavf] error reading packet: Resource temporarily unavailable.
[lavf] error reading packet: Resource temporarily unavailable.
[lavf] error reading packet: Resource temporarily unavailable.
[lavf] error reading packet: Resource temporarily unavailable.
[lavf] error reading packet: Resource temporarily unavailable.
[lavf] error reading packet: Resource temporarily unavailable.
[lavf] error reading packet: Resource temporarily unavailable.
[lavf] error reading packet: Resource temporarily unavailable.
[lavf] error reading packet: Resource temporarily unavailable.
[lavf] error reading packet: Resource temporarily unavailable.
[lavf] ...treating it as fatal error.
```
Runs for a split second and then closes. 

```
demuxer-force-retry-on-eof=yes
```
Keeps it going, though choppy and randomly drops later. Will keep playing with options. Thanks!

**Comment by devsifi** [comment]
I've had more luck piping through FFmpeg to MPV then using MPV directly, however latency is slightly worse than Genki Arcade itself
```
ffmpeg -f avfoundation \
        -video_size 1280x720 -framerate 60 -i "ShadowCast" \
        -vcodec libx264 -preset ultrafast -tune zerolatency -b:v 5000k\
        -f matroska - | mpv --no-cache --untimed --no-demuxer-thread --vd-lavc-threads=1 -
```

**Comment by emilyst** [comment]
I've been trying for several days to see if I can get this to work. I have found this issue also occurs in IINA (obviously) but also even VLC and a few other media applications. It's as if they know how to grab the first available frame, but then something happens and they can't fetch the rest.

Based on "Resource temporarily unavailable" it seems as if `libavformat` is hitting an `EAGAIN` error reading the packets because it's requesting a non-blocking read from a device which necessarily blocks.

**Comment by ikester** [comment]
> Based on "Resource temporarily unavailable" it seems as if `libavformat` is hitting an `EAGAIN` error reading the packets because it's requesting a non-blocking read from a device which necessarily blocks.

That's interesting. I don't know enough about `libavformat` but I would imagine `ffmpeg` would use the same call path. So it's puzzling that opening the stream with `ffmpeg` like it was suggested above **does** work. For example, this works for me with the MBP FaceTime camera:

```
ffmpeg -f avfoundation -video_size 1920x1080 -framerate 30 -pixel_format uyvy422 -i "0:none" -f matroska \
 - | mpv -profile=low-latency --untimed --fs -
```

**Comment by emilyst** [comment]
> > Based on "Resource temporarily unavailable" it seems as if `libavformat` is hitting an `EAGAIN` error reading the packets because it's requesting a non-blocking read from a device which necessarily blocks.
> 
> That's interesting. I don't know enough about `libavformat` but I would imagine `ffmpeg` would use the same call path. So it's puzzling that opening the stream with `ffmpeg` like it was suggested above **does** work. For example, this works for me with the MBP FaceTime camera:
> 
> ```
> ffmpeg -f avfoundation -video_size 1920x1080 -framerate 30 -pixel_format uyvy422 -i "0:none" -f matroska \
>  - | mpv -profile=low-latency --untimed --fs -
> ```

Yeah, I agree it’s a bit odd. But `ffmpeg` seems to struggle to capture the stream smoothly as well. I would almost guess it’s just opening and closing the stream repeatedly, based on how the picture and audio stutter when I try it.

**Comment by ikester** [comment]
> > > Based on "Resource temporarily unavailable" it seems as if `libavformat` is hitting an `EAGAIN` error reading the packets because it's requesting a non-blocking read from a device which necessarily blocks.
> > 
> > 
> > That's interesting. I don't know enough about `libavformat` but I would imagine `ffmpeg` would use the same call path. So it's puzzling that opening the stream with `ffmpeg` like it was suggested above **does** work. For example, this works for me with the MBP FaceTime camera:
> > ```
> > ffmpeg -f avfoundation -video_size 1920x1080 -framerate 30 -pixel_format uyvy422 -i "0:none" -f matroska \
> >  - | mpv -profile=low-latency --untimed --fs -
> > ```
> 
> Yeah, I agree it’s a bit odd. But `ffmpeg` seems to struggle to capture the stream smoothly as well. I would almost guess it’s just opening and closing the stream repeatedly, based on how the picture and audio stutter when I try it.

I don't see `ffmpeg` struggle on a MacBook Pro M1 Max, admittedly one of their higher end machines. It only slows down for me (drops frames, lags) when I add non-accelerated filters to the pipeline, which is what I'm trying to use it for.

In any event, I would love to find out why `mpv` is hitting the `EAGAIN` error when `ffmpeg` doesn't. Or if `fmpeg` does get it, how does it manage to ignore it or recover from it.

**Comment by maxexcloo** [comment]
Did anyone work this out? I'm having trouble with playing the stream with MPV also...
